In [1]:
import pandas as pd

In [95]:
loans_2007 = pd.read_csv("loans_2007.csv")
loans_2007_copy = pd.read_csv("loans_2007.csv")

C:\Users\ASUS\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2728: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [26]:
loans_2007.shape

(42538, 52)

In [4]:
loans_2007.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42538 entries, 0 to 42537
Data columns (total 52 columns):
id                            42538 non-null object
member_id                     42535 non-null float64
loan_amnt                     42535 non-null float64
funded_amnt                   42535 non-null float64
funded_amnt_inv               42535 non-null float64
term                          42535 non-null object
int_rate                      42535 non-null object
installment                   42535 non-null float64
grade                         42535 non-null object
sub_grade                     42535 non-null object
emp_title                     39909 non-null object
emp_length                    41423 non-null object
home_ownership                42535 non-null object
annual_inc                    42531 non-null float64
verification_status           42535 non-null object
issue_d                       42535 non-null object
loan_status                   42535 non-null object
p

In [5]:
import matplotlib.pyplot as plt
%matplotlib inline

In [10]:
import numpy as np

In [31]:
loans_2007_copy.iloc[:3,:18]

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501  1296599.0     5000.0       5000.0           4975.0   36 months   
1  1077430  1314167.0     2500.0       2500.0           2500.0   60 months   
2  1077175  1313524.0     2400.0       2400.0           2400.0   36 months   

  int_rate  installment grade sub_grade emp_title emp_length home_ownership  \
0   10.65%       162.87     B        B2       NaN  10+ years           RENT   
1   15.27%        59.83     C        C4     Ryder   < 1 year           RENT   
2   15.96%        84.33     C        C5       NaN  10+ years           RENT   

   annual_inc verification_status   issue_d  loan_status pymnt_plan  
0     24000.0            Verified  Dec-2011   Fully Paid          n  
1     30000.0     Source Verified  Dec-2011  Charged Off          n  
2     12252.0        Not Verified  Dec-2011   Fully Paid          n

## As it is a huge dataset containing 54 columns,so we are observing the first 18 columns

In [96]:
loans_2007 = loans_2007.drop(["id", "member_id", "funded_amnt", "funded_amnt_inv", "grade", "sub_grade", "emp_title", "issue_d"], axis=1)

## I have removed some columns from the dataset because it was leaking information about the future i.e. after the loan is being paid off.

In [20]:
loans_2007_copy.iloc[:3,18:37]

purpose                 title zip_code addr_state    dti  \
0     credit_card              Computer    860xx         AZ  27.65   
1             car                  bike    309xx         GA   1.00   
2  small_business  real estate business    606xx         IL   8.72   

   delinq_2yrs earliest_cr_line  inq_last_6mths  open_acc  pub_rec  revol_bal  \
0          0.0         Jan-1985             1.0       3.0      0.0    13648.0   
1          0.0         Apr-1999             5.0       3.0      0.0     1687.0   
2          0.0         Nov-2001             2.0       2.0      0.0     2956.0   

  revol_util  total_acc initial_list_status  out_prncp  out_prncp_inv  \
0      83.7%        9.0                   f        0.0            0.0   
1       9.4%        4.0                   f        0.0            0.0   
2      98.5%       10.0                   f        0.0            0.0   

   total_pymnt  total_pymnt_inv  total_rec_prncp  
0  5863.155187          5833.84          5000.00  
1  1008.710000          1008.71           456.46  
2  3005.666844          3005.67          2400.00

## Lets observe the next set of 18 columns

In [97]:
loans_2007 = loans_2007.drop(["zip_code", "out_prncp", "out_prncp_inv", "total_pymnt", "total_pymnt_inv", "total_rec_prncp"], axis=1)

## Again i removed those columns that were leaking data from the future

In [21]:
loans_2007_copy.iloc[:3,37:]

total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0         863.16                 0.0        0.00                     0.00   
1         435.17                 0.0      117.08                     1.11   
2         605.67                 0.0        0.00                     0.00   

  last_pymnt_d  last_pymnt_amnt last_credit_pull_d  \
0     Jan-2015           171.62           Jun-2016   
1     Apr-2013           119.66           Sep-2013   
2     Jun-2014           649.91           Jun-2016   

   collections_12_mths_ex_med  policy_code application_type  acc_now_delinq  \
0                         0.0          1.0       INDIVIDUAL             0.0   
1                         0.0          1.0       INDIVIDUAL             0.0   
2                         0.0          1.0       INDIVIDUAL             0.0   

   chargeoff_within_12_mths  delinq_amnt  pub_rec_bankruptcies  tax_liens  
0                       0.0          0.0                   0.0        0.0  
1                       0.0          0.0                   0.0        0.0  
2                       0.0          0.0                   0.0        0.0

## Observing the last 18 columns of the dataset

In [98]:
loans_2007 = loans_2007.drop(["total_rec_int", "total_rec_late_fee", "recoveries", "collection_recovery_fee", "last_pymnt_d", "last_pymnt_amnt"], axis=1)

## The columns which were leaking data about the future were removed

In [30]:
loans_2007.shape

(42538, 32)

## So we removed about 20 columns from the dataset because they were leaking information about future.

In [32]:
loans_2007["loan_status"].unique()

array(['Fully Paid', 'Charged Off', 'Current', 'In Grace Period',
       'Late (31-120 days)', 'Late (16-30 days)', 'Default', nan,
       'Does not meet the credit policy. Status:Fully Paid',
       'Does not meet the credit policy. Status:Charged Off'],
      dtype=object)

## Here we will consider a binary classification by considering "Fully Paid" and "Charged Off" option only

In [99]:
loans_2007 = loans_2007.loc[(loans_2007["loan_status"] == "Fully Paid") | (loans_2007["loan_status"] == "Charged Off")]

In [100]:
loans_2007["loan_status"] = loans_2007["loan_status"].map({"Fully Paid":1,"Charged Off":0})

In [37]:
loans_2007["loan_status"].unique()

array([1, 0], dtype=int64)

In [101]:
cols_drop = []
for col in loans_2007.columns:
    no_nan = loans_2007[col].dropna()
    if len(no_nan.unique()) == 1:
        cols_drop.append(col)
loans_2007 = loans_2007.drop(cols_drop,axis = 1)   

## The above code is done to remove those columns which has only one unique value

In [41]:
loans_2007.shape

(38770, 23)

## Nine columns were removed from the dataset that contained only one unique value

In [42]:
loans_2007.isnull().sum()

loan_amnt                  0
term                       0
int_rate                   0
installment                0
emp_length              1036
home_ownership             0
annual_inc                 0
verification_status        0
loan_status                0
purpose                    0
title                     11
addr_state                 0
dti                        0
delinq_2yrs                0
earliest_cr_line           0
inq_last_6mths             0
open_acc                   0
pub_rec                    0
revol_bal                  0
revol_util                50
total_acc                  0
last_credit_pull_d         2
pub_rec_bankruptcies     697
dtype: int64

In [11]:
import numpy as np

In [111]:
loans_2007["emp_length"] = loans_2007["emp_length"].map({
        "10+ years": 10,
        "9 years": 9,
        "8 years": 8,
        "7 years": 7,
        "6 years": 6,
        "5 years": 5,
        "4 years": 4,
        "3 years": 3,
        "2 years": 2,
        "1 year": 1,
        "< 1 year": 0,
        "n/a": 0
    })

## i changed the datatype of "emp_length" column from object to float

In [75]:
loans_2007.loc[loans_2007["emp_length"] == 4,"emp_length"] = 5

## Here i experimented with the null values of the "emp_length" column by filling it up with the median of the series

In [102]:
loans_2007 = loans_2007.drop("pub_rec_bankruptcies", axis=1)
loans_2007 = loans_2007.dropna(axis=0)
print(loans_2007.dtypes.value_counts())

object     11
float64    10
int64       1
dtype: int64


## The column which had more than 1% of the data size as null values were removed

In [57]:
object_columns_df = loans_2007.select_dtypes(include=["object"])
print(object_columns_df.iloc[0])

term                     36 months
int_rate                    10.65%
home_ownership                RENT
verification_status       Verified
purpose                credit_card
title                     Computer
addr_state                      AZ
earliest_cr_line          Jan-1985
revol_util                   83.7%
last_credit_pull_d        Jun-2016
Name: 0, dtype: object


## Observing the object datatype columns

In [59]:
cols = ['home_ownership', 'verification_status', 'term', 'addr_state']
for c in cols:
    print(loans_2007[c].value_counts())

RENT        18513
MORTGAGE    17111
OWN          2984
OTHER          96
NONE            3
Name: home_ownership, dtype: int64
Not Verified       16696
Verified           12289
Source Verified     9722
Name: verification_status, dtype: int64
 36 months    29040
 60 months     9667
Name: term, dtype: int64
CA    6958
NY    3713
FL    2791
TX    2667
NJ    1798
IL    1483
PA    1473
VA    1376
GA    1364
MA    1301
OH    1179
MD    1026
AZ     850
WA     822
CO     770
NC     753
CT     730
MI     712
MO     670
MN     603
NV     481
SC     462
WI     441
AL     437
OR     436
LA     430
KY     315
OK     290
KS     260
UT     254
AR     237
DC     209
RI     196
NM     184
WV     172
NH     166
HI     166
DE     113
MT      83
WY      80
AK      78
SD      61
VT      53
MS      19
TN      17
IN       9
ID       6
NE       5
IA       5
ME       3
Name: addr_state, dtype: int64


## The home_ownership, verification_status and term columns each contain a few discrete categorical values. We should encode these columns as dummy variables and keep them.
## Lastly, the addr_state column contains many discrete values and we'd need to add 49 dummy variable columns to use it for classification. This would make our Dataframe much larger and could slow down how quickly the code runs. Let's remove this column from consideration.

In [51]:
print(loans_2007["purpose"].value_counts())
print(loans_2007["title"].value_counts())

debt_consolidation    18102
credit_card            5036
other                  3856
home_improvement       2894
major_purchase         2149
small_business         1754
car                    1510
wedding                 923
medical                 680
moving                  576
vacation                375
house                   367
educational             317
renewable_energy        101
Name: purpose, dtype: int64
Debt Consolidation                              2102
Debt Consolidation Loan                         1628
Personal Loan                                    640
Consolidation                                    493
debt consolidation                               485
Credit Card Consolidation                        353
Home Improvement                                 345
Debt consolidation                               324
Small Business Loan                              309
Credit Card Loan                                 305
Personal                                         3

## It seems like the purpose and title columns do contain overlapping information but we'll keep the purpose column since it contains a few discrete values¶

In [103]:
loans_2007 = loans_2007.drop(["last_credit_pull_d", "earliest_cr_line", "addr_state", "title"], axis=1)

In [104]:
loans_2007["int_rate"] = loans_2007["int_rate"].str.rstrip("%").astype("float")
loans_2007["revol_util"] = loans_2007["revol_util"].str.rstrip("%").astype("float")

## We converted the "int_rate" and "revol_util" to float type

In [105]:
cat_columns = ["home_ownership", "verification_status", "purpose", "term"]
dummy_df = pd.get_dummies(loans_2007[cat_columns])
loans_2007 = pd.concat([loans_2007, dummy_df], axis=1)
loans_2007 = loans_2007.drop(cat_columns, axis=1)

In [107]:
loans_2007.shape

(37675, 38)

## The dummies were created for the relevant columns and the original columns were subsequently dropped

In [63]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()

In [108]:
features = loans_2007.columns.drop("loan_status")

In [109]:
target = "loan_status"

In [112]:
lr.fit(loans_2007[features],loans_2007[target])

LogisticRegression(C=1.0, class_weight={0: 12, 1: 1}, dual=False,
          fit_intercept=True, intercept_scaling=1, max_iter=100,
          multi_class='ovr', n_jobs=1, penalty='l2', random_state=None,
          solver='liblinear', tol=0.0001, verbose=0, warm_start=False)

In [113]:
predictions = lr.predict(loans_2007[features])

In [68]:
from sklearn.model_selection import cross_val_predict

In [114]:
predictions = cross_val_predict(lr, loans_2007[features],loans_2007[target], cv=3)

In [115]:
predictions

array([0, 0, 0, ..., 1, 1, 0], dtype=int64)

In [116]:
predictions = pd.Series(predictions)
# False positives.
fp_filter = (predictions == 1) & (loans_2007["loan_status"] == 0)
fp = len(predictions[fp_filter])

# True positives.
tp_filter = (predictions == 1) & (loans_2007["loan_status"] == 1)
tp = len(predictions[tp_filter])

# False negatives.
fn_filter = (predictions == 0) & (loans_2007["loan_status"] == 1)
fn = len(predictions[fn_filter])

# True negatives
tn_filter = (predictions == 0) & (loans_2007["loan_status"] == 0)
tn = len(predictions[tn_filter])
# Rates
tpr = tp  / (tp + fn)
fpr = fp  / (fp + tn)
print(tpr)
print(fpr)

0.13093209229155742
0.12754195459032577


In [117]:
lr = LogisticRegression(class_weight="balanced")
predictions = cross_val_predict(lr, loans_2007[features],loans_2007[target], cv=3)
predictions = pd.Series(predictions)

# False positives.
fp_filter = (predictions == 1) & (loans_2007["loan_status"] == 0)
fp = len(predictions[fp_filter])

# True positives.
tp_filter = (predictions == 1) & (loans_2007["loan_status"] == 1)
tp = len(predictions[tp_filter])

# False negatives.
fn_filter = (predictions == 0) & (loans_2007["loan_status"] == 1)
fn = len(predictions[fn_filter])

# True negatives
tn_filter = (predictions == 0) & (loans_2007["loan_status"] == 0)
tn = len(predictions[tn_filter])

# Rates
tpr = tp / (tp + fn)
fpr = fp / (fp + tn)

print(tpr)
print(fpr)

0.6288345568956476
0.6159921026653504


In [126]:
penalty = {
    0: 12,
    1: 1}
lr = LogisticRegression(class_weight  =penalty)
predictions = cross_val_predict(lr, loans_2007[features],loans_2007[target], cv=5)
predictions = pd.Series(predictions)

# False positives.
fp_filter = (predictions == 1) & (loans_2007["loan_status"] == 0)
fp = len(predictions[fp_filter])

# True positives.
tp_filter = (predictions == 1) & (loans_2007["loan_status"] == 1)
tp = len(predictions[tp_filter])

# False negatives.
fn_filter = (predictions == 0) & (loans_2007["loan_status"] == 1)
fn = len(predictions[fn_filter])

# True negatives
tn_filter = (predictions == 0) & (loans_2007["loan_status"] == 0)
tn = len(predictions[tn_filter])

# Rates
tpr = tp / (tp + fn)
fpr = fp / (fp + tn)

print(tpr)
print(fpr)

0.1323741478762454
0.1362290227048371


## We want to minimize the false positive rate.So it will predict the bad loans but we will be losing on opportunities as well.

In [127]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.cross_validation import cross_val_predict
penalty = {
    0: 12,
    1: 1}
rf = RandomForestClassifier(class_weight=penalty, random_state=1)
predictions = cross_val_predict(rf, loans_2007[features],loans_2007[target], cv=5)
predictions = pd.Series(predictions)

# False positives.
fp_filter = (predictions == 1) & (loans_2007["loan_status"] == 0)
fp = len(predictions[fp_filter])

# True positives.
tp_filter = (predictions == 1) & (loans_2007["loan_status"] == 1)
tp = len(predictions[tp_filter])

# False negatives.
fn_filter = (predictions == 0) & (loans_2007["loan_status"] == 1)
fn = len(predictions[fn_filter])

# True negatives
tn_filter = (predictions == 0) & (loans_2007["loan_status"] == 0)
tn = len(predictions[tn_filter])

# Rates
tpr = tp / (tp + fn)
fpr = fp / (fp + tn)

print(tpr)
print(fpr)

0.9712244362873623
0.9727541954590325


##  For a conservative investor, this means that they make money as long as the interest rate is high enough to offset the losses from 11% of borrowers defaulting, and that the pool of 13% of borrowers is large enough to make enough interest money to offset the losses.